sunkit_image/granule.py

```diff
@@ -0,0 +1,331 @@
+"""
+This module contains functions that will segment images for granule detection.
+"""
+import logging
+
+import numpy as np
+import scipy.ndimage as sndi
+import skimage
+from sklearn.cluster import KMeans
+
+import sunpy
+import sunpy.map
+
+__all__ = [
+    "segment",
+    "get_threshold",
+    "trim_intergranules",
+    "mark_faculae",
+    "kmeans_segment",
+    "cross_correlation",
+]
+
+
+def segment(data_map, skimage_method, res, mark_dim_centers=False):
+    """
+    Segment optical image of the solar photosphere into tri-value maps
+    with 0 = intergranule, 0.5 = faculae, 1 = granule.
+
+    Parameters
+    ----------
+    data_map : `SunPy map`
+        SunPy map containing data to segment
+    skimage_method : `str`
+        skimage thresholding method, with options 'otsu', 'li', 'isodata',
+        'mean', 'minimum', 'yen', and 'triangle'.
+        Note - depending on input data, one or more of these methods may be
+        signifcantly better or worse than the others. Typically, 'li', 'otsu',
+        'mean', and 'isodata' are good choices, 'yen' and 'triangle' over-
+        identify intergranule material, and 'minimum' overidentifies granules.
+    res : `float`
+        Spatial resolution (arcsec/pixel) of the data
+    mark_dim_centers : `bool`
+        Whether to mark dim granule centers as a seperate catagory for future
+        exploration.
+
+    Returns
+    -------
+    segmented_map : `SunPy map`
+        SunPy map containing segmentated image (with the original header)
+    """
+
+    if not isinstance(data_map, sunpy.map.mapbase.GenericMap):
+        raise TypeError("Input must be sunpy map.")
+
+    methods = ["otsu", "li", "isodata", "mean", "minimum", "yen", "triangle"]
+    if skimage_method not in methods:
+        raise TypeError("Method must be one of: " + str(methods))
+
+    data = data_map.data
+    header = data_map.meta
+
+    # apply median filter
+    median_filtered = sndi.median_filter(data, size=3)
+
+    # apply threshold
+    threshold = get_threshold(median_filtered, skimage_method)
+
+    # initial skimage segmentation

```

Comment:
What does skimage have to do with this line?

Reply:
I suppose in my mind "# initial skimage segmentation" just indicates that the line `segmented_image = np.uint8(median_filtered > threshold)` applies the skimage threshold to the data. I will change this to be clearer, so that line 65 is "# get initial skimage threshold" and line 68 is "# apply threshold"

Reply:
That makes sense.